### ISSUE TITLE
Automate publishing process via CI

### ISSUE THREAD

**Issue by mtrezza** [parse-community/Parse-SDK-Flutter#668]
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/Parse-SDK-Flutter/security/policy).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/main/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/Parse-SDK-Flutter/issues?q=is%3Aissue).

### Current Limitation
Currently, publishing a new version of the Parse Flutter SDK on pub.dev is done manually. This poses several issues:
- Inefficient, as manual steps are required
- Possible divergence of GH versions and pub.dev versions
- Separate permissions management, instead of managing publishing permission as part of GH team permissions

### Feature / Enhancement Description
Automate publishing process via GitHub CLI, analogous to, for example publishing parse-server on npm.

TODOs:
- [x] Add auto-publishing to pub.dev (triggered on GitHub release) https://github.com/parse-community/Parse-SDK-Flutter/pull/719
- [x] Validate by creating a manual release (3.1.x or 3.2.0) **-- version tag without `V` prefix!**
- [x] Add auto-release workflow (branch renaming, changelog-generation, etc) https://github.com/parse-community/Parse-SDK-Flutter/pull/1049
- [ ] Validate by creating a automatic release via PR-merge
  - [ ] Flutter release is incorrectly triggered despite only dependency upgrade in dart package
  - [ ] Packages are not published to pub dir

**Comment by mtrezza** [comment]
@RodrigoSMarques It would be great if you could have a look into this, since you seem to be most familiar with the pub.dev publishing process. I am of course available to assist from the GitHub CI side, to we can automate this. I think a good starting point would be to look at the parse-server workflows, specifically, the [release](https://github.com/parse-community/parse-server/tree/master/.github/workflows) flow.

**Comment by fischerscode** [comment]
@mtrezza @RodrigoSMarques 
If you want, I could have a look at this, too.

**Comment by fischerscode** [comment]
Some more information as I'm now home:

Currently there is no recommended way to publish dart/Flutter packages to pub.dev automatically. The community uses the access and refresh token `pub publish` created at the first publish. Technically those tokens are not package but account bound and belong to the personal account. This meas, although they work for publishing, in theory they are a security risk, as someone with push access to this repo could extract the (GitHub) Secret containing those tokens and use them in order to authenticate as the user who created them.
As an example: If @RodrigoSMarques would create the GitHub Secrets using his personal account, anybody who has push access to this repo **could** steal those and publish to his package.
Please note: This is just theory and I don't think anybody would do such a thing (right?) esp. as the fake release could be easily reverted by the pub.dev team. But never the less, it is worth considering the potential risk.

The options I see:
1. Accept the risk and use the [this](https://github.com/marketplace/actions/dart-and-flutter-package-publisher) action. (Publish CI as soon as possible)
2. Create a Google Account that has just access to those packages. (Nothing relevant could be stolen.)
3. Publish manually until an official way to publish using CI is added. [See this issue](https://github.com/dart-lang/pub/issues/2227) (Personally I am using this way for a small package.)

Some thinks the automation should do:
- On pre-release: make sure to only [publish prereleases](https://dart.dev/tools/pub/publishing#publishing-prereleases). (Check version)
- On release: makes sure not to publish as a prerelease. (Check version)
- Release the dart package first and the flutter package second (might fail otherwise as flutter might/should depend on dart).

**Comment by mtrezza** [comment]
@fischerscode Thanks for the analysis.

Yes, for now we'll publish manually until we found a feasible way to automate this.

We should do a careful risk analysis regarding automated publishing and permission exposure. The risks is not only malicious intent by a single person, but maybe more accidental actions or compromised 3rd party accounts, because we cannot control how someone stores their own account credentials.

**Comment by Mabenan** [comment]
Maybe i can give some new information on stealing secrets from github actions as i already tried this myself to get a lost key back. First of it is not so easy as every secret used in github actions can not be simply printed. I only could get by echo it to a txt file and upload this text file somewhere. Maybe we there are ways of securing the access to the secrets by restricting the environment to branches (or maybe cleanup and restrict the branch creation all together) and then restrict the merging to these branches.

**Comment by mtrezza** [comment]
We already use GitHub secrets for NPM and docker publishing token in other repos; we can assume the feature is safe as that is exactly what it's designed for. As I understand it, secrets are only applied on the original repo branches, not on fork branches or PRs. We just have to be careful about what we merge, but that we have to be anyway.

**Comment by RodrigoSMarques** [comment]
Hi @mtrezza.

How can I contact you so we can talk about releasing new versions of the package?

Since Phil left the project we haven't published any updates, because I don't know how the process was defined.
 
I believe that as the package is now publisher verified, you could invite me to publish the version.

I await your contact.

**Comment by mtrezza** [comment]
Maybe we can already set up the CI to automate the publishing in the future? Do you want to look into that and open a PR for a GitHub workflow?

**Comment by RodrigoSMarques** [comment]
Hi @mtrezza . 
I was away for vacation.

I don't know if there is an option for automated deployment.

My sugestion. Do the update manually and then try to automate.

We have several bugs to be updated and as I have other projects, time is limited to check before the update.

My sugestion. Do the update manually and then try to automate.

We have several bugs to be updated and as I have other projects, time is limited to check before the update.

In this issue we have the analysis about the deployment automation.

**Comment by mtrezza** [comment]
Sure, @RodrigoSMarques I think we could give you temporary maintainer access if you would be willing to publish a release?

**Comment by mtrezza** [comment]
@RodrigoSMarques Friendly ping in case you overlooked my previous comment.

**Comment by mtrezza** [comment]
I should be able to give you permissions later today.

**Comment by mtrezza** [comment]
@RodrigoSMarques you've been granted publisher permission, please let me know if it works and thanks for your help to publish manually.

To hopefully gather more interest in automating publishing, I've added a bounty label; @fischerscode you wrote a comprehensive analysis, would you want to take look and open a PR?

**Comment by mtrezza** [comment]
Amazing @fischerscode, I would prefer to add auto-release to this SDK as well, just as we have with for example the Parse JS SDK. That means on each commit (depending on the commit message), a release will be automatically published.

The auto-release mechanism is essentially the same as in the Parse JS SDK, but it comes with its own complexities, so I would break it into two PRs:
1. Automate pub.dev publishing when a new GH release is published (we still need to manually add a changelog entry with every PR and determine the next version number based on the changelog entries)
2. Move to auto-release (if you are interested you can do that too, otherwise I can take over this part)

**Comment by fischerscode** [comment]
@mtrezza I've created PR #719. It addresses the first point.
Pub.dev does not support token authentication at the moment, as a result, the action requires the refresh and access token (OAuth2) of a Google account, that can publish. I would suggest creating a bot account, granting it publish rights (pub.dev), use it to log in to google using pub and then create the secrets `PUBLISH_OAUTH_ACCESS_TOKEN`, `PUBLISH_OAUTH_REFRESH_TOKEN` based on `credentials.json`.

https://github.com/k-paxian/dart-package-publisher#accesstoken

I assume you are one of the administrators of the [publisher](https://pub.dev/publishers/parseplatform.org/packages). So I think it would be best, if you would create the bot account.

**Comment by fischerscode** [comment]
> 2\. if you are interested you can do that too  based on how we already do it in the Parse JS SDK and I'll be on standby for any questions

I could do that, but (as you wrote) we should focus on Nr. 1 first.

**Comment by mtrezza** [comment]
@fischerscode That's amazing and amazingly quick! I will create a Google account for that and add the secrets so maybe we can even try that out with our first publishing. 

@RodrigoSMarques You may want to wait with manual publishing, so we try this out?

**Comment by RodrigoSMarques** [comment]
> @fischerscode That's amazing and amazingly quick! I will create a Google account for that and add the secrets so maybe we can even try that out with our next publishing.
> 
> @RodrigoSMarques You may want to wait with manual publishing, so we try this out?

We can wait to test.

Who was having problems, now has a version to continue through the Fork

**Comment by mtrezza** [comment]
Do I understand it correctly, that these 2 packages (dart, flutter) always will have the same version number?

**Comment by fischerscode** [comment]
For now, they always had.
But that's just how we did it till now.

We could treat both as independent.
(Dart package goes through the release cycle and the flutter one just uses the dart one as a dependency. Similar to dart vs flutter (The software projects.))

**Comment by mtrezza** [comment]
Is that dependency opaque to the developer or do they have to manually choose the version number of the dart dependency when using the flutter sdk?

**Comment by fischerscode** [comment]
Currently, they can just add `parse_server_sdk_flutter: ^3.1.0`. `parse_server_sdk: ^3.1.0` is then automatically added, since it is a dependency of `parse_server_sdk_flutter`.

I we would release the packages independently (different versions) the developer would still be able to just add parse_server_sdk_flutter. But which version of parse_server_sdk is actually used might be a bit less obvious, but still, it could be checked by using the cli tools or checking `pubspec.lock`.

**Comment by mtrezza** [comment]
Got it, I'm asking for release automation so we know our constraints when designing the mechanism.

**Comment by fischerscode** [comment]
[This](https://pub.dev/packages/melos) might be the tool we are searching for the auto-release workflow.

**Comment by fischerscode** [comment]
Just looked a little bit more into melos... It's definitely the tool for us.
It even updates version/changelog based on git commits.
I might go over some of the testing suff and use it, since it handles the local path automatically and therefore makes the actions easier to read/maintain.

...but that's a project for tomorrow.

**Comment by mtrezza** [comment]
If possible I'd prefer to use semantic release, it also supports monorepo. It is the tool we use for all other repos.

**Comment by fischerscode** [comment]
Ok, but does it support dart/flutter? Otherwise we might have to write a script that updates the versions in `pubspec.yaml`.

**Comment by mtrezza** [comment]
It's quite versatile, so I suppose it will work here as well. For example, we use it in the ObjC and Android SDKs.

**Comment by mtrezza** [comment]
Can we try to publish a release?

**Comment by RodrigoSMarques** [comment]
Not. I haven't had time to finish the updates yet. I'll let you know as soon as I'm done.

**Comment by mtrezza** [comment]
OK, I guess we could also make a quick patch release 3.1.1 for the Android example app fix in the meantime. We need a couple of releases to test this out anyway, so we don't have to wait until we get more PRs merged.

**Comment by fischerscode** [comment]
> OK, I guess we could also make a quick patch release 3.1.1 for the Android example app fix in the meantime. We need a couple of releases to test this out anyway, so we don't have to wait until we get more PRs merged.

Sounds like a plan. Maybe make it a pre-release? It works the identically and we would not spam the releases.

**Comment by mtrezza** [comment]
I am hesitant to make any pre-releases on the `master` branch. That would violate our later branch model and could cause issues with release automation. Let's just do a simple stable release where the example app is fixed. If we can get that to work then we should be good for stable releases. Pre-releases probably won't be done until we have release-automation.

**Comment by fischerscode** [comment]
> Update README

Document that we are using ci?

> We'll also switch the version syntax from `V#.#.#` to `#.#.#`

We should create `#.#.#` tags that match the `V#.#.#` ones due to consistency.

> What do you think?

Sounds good.

**Comment by mtrezza** [comment]
Continuing discussion from https://github.com/parse-community/Parse-SDK-Flutter/pull/736#issuecomment-1114368790:

It seems weird that both changelogs are identical. There may be a change in the Flutter package that is irrelevant for the dart package but would appear in the dart changelog. I am beginning to doubt whether we should really keep the versions in sync.

**Comment by fischerscode** [comment]
> There may be a change in the Flutter package that is irrelevant for the dart package but would appear in the dart changelog.

Since the split, there has been no flutter-only change. I case there would have been one, it would only have been added to the flutter changelog.

> I am beginning to doubt whether we should really keep the versions in sync.

I think we should stop doing that, the approach [flutterfire](https://firebase.flutter.dev/) uses is better, it think.

> If we maintain the versions independent from each other, do we have to change anything in the release workflow that we already merged?

dry-run would not make sens any more, since the flutter package could be updated independent of the dart package and vice versa. As a result we would only publish one. Due to the version not changing, dry-run would fail for the other one.

**Comment by mtrezza** [comment]
> We would have to remember to ask flutter developers to add the used version of the dart package to new issues.

I thought we can infer that from the flutter version, since the dart dependency is fixed per version?

**Comment by fischerscode** [comment]
If we specify the dart dependency with ^x.x.x, pub will update to non major versions.

**Comment by mtrezza** [comment]
Maybe we shouldn't do that, at least for our own dart dependency, since it has to be compatible with the flutter package.

**Comment by fischerscode** [comment]
So you mean:
- publish dart package if there are changes
- hard link flutter package to the dart package
- publish flutter package when it has changes or the version of the dart package changes

**Comment by mtrezza** [comment]
I look at them as separate packages. If there is any change in dart, then a new dart version is published. That doesn't mean that a new flutter version is published at the same time. 

You could think of the 2 package being from different maintainers that don't have to work in sync. However, since it's a monorepo, we have the advantage that a single PR can can make changes in both packages and run the tests in one go.

**Comment by fischerscode** [comment]
> I look at them as separate packages.

That's what I would prefer as well. And then the flutter package should depend on the dart package using "^".

**Comment by mtrezza** [comment]
Why?

**Comment by fischerscode** [comment]
It ^3.1.0 is the short form for ">=3.1.0 <4.0.0" so we would not have to publish the flutter one, every time the dart one changes (the dart one has much more code, so changes there will occur more frequently).

**Comment by mtrezza** [comment]
I think I follow you now, you mean that there are more changes in dart than in flutter, so we would have to publish a new flutter version practically every time we publish a new dart version, right? So if we use a range we don't have to do that and we know flutter will always be compatible with dart if we follow semver.

Could you explain again when the lock file (and therefore the dart dependency in flutter) is updated? Is that only when we publish the package to the registry, or also at some point when the developer uses the package?

**Comment by fischerscode** [comment]
> I think I follow you now, you mean that there are more changes in dart than in flutter, so we would have to publish a new flutter version practically every time we publish a new dart version, right? So if we use a range we don't have to do that and we know flutter will always be compatible with dart if we follow semver.

Correct 👍 

> Could you explain again when the lock file (and therefore the dart dependency in flutter) could change? Is that only when we publish the package to the registry, or also at some point when the developer uses the package?

We do neither version control the lock file nor do we publish it to pub.dev.

The pubspec.lock file contains the actual used versions (no range) of the dependencies and where they are stored.
A developer working an application would version control this file to make sure the entire team works with the same dependency versions.
When we publish a new version of the dart package (minor or patch), the developer of an flutter application (using parse_server_sdk_flutter) would not use that new version immediately. But once running `flutter pub upgrade` the version of `parse_server_sdk` would update in the lock file while `parse_server_sdk_flutter` would stay the same.

**Comment by mtrezza** [comment]
> We do neither version control the lock file nor do we publish it to pub.dev.

I'm not sure about this approach. When we publish a version that has been CI tested, we basically say to developers: "this is a version that we checked, is stable and you can upgrade without issues." Is it considered a stable snapshot that is reproducible for everyone at any time, without any changes in code.

We cannot say that if dependency versions are not fixed. In my view (and as I believe as is commonly understood in the ecosystem) it is not the developer's responsibility to choose the dependency versions and then run the CI again to ensure all is working. It is the maintainers (our) responsibility.

Therefore we need to commit the lock file, which is the snapshot that passes the CI. We can keep the range operators in the definitions to mKe life easier for us since there is no 3rd party service (like Snyk for npm) that upgrades for us.

**Comment by mtrezza** [comment]
@fischerscode was my previous comment convincing? 🙂

**Comment by fischerscode** [comment]
> @fischerscode @RodrigoSMarques was my previous comment convincing? 🙂

Sorry, missed that comment...

> Therefore we need to commit the lock file, which is the snapshot that passes the CI.

`pub publish` wont upload the lock file. So we can't publish it.
Committing it to git will have no effect, since almost nobody will include the package using a git dependency.

> We can keep the range operators in the definitions to make life easier for us since there is no 3rd party service (like Snyk for npm) that upgrades dependencies for us.

The advantage of not pinning the versions and instead allowing non breaking updates (using `^`) appears, when a dependency gets patched. The developers could already use that version without a new version of the parse package being published. Additionally, it is very annoying (as a developer) when a package gets a new feature and updates the minimal dependencies, but you can't use it because some other package doesn't allow newer the newer versions of those dependencies.

> In my view (and I believe as is commonly understood in the ecosystem) it is not the developer's responsibility to choose the dependency versions

As far as I know, every package/plugin published by the core flutter team uses the range operators. https://github.com/flutter/plugins

To verify that they really do not stick the versions:
- `flutter create dependency_test` (Create a new flutter project)
- `cd dependency_test`
- `flutter pub add camera:0.9.4+21` (Add specific version of the camera plugin)

Pubspec.lock will show, that `version: "3.1.0"` of `quiver` is used, even through / because the [pubspec.yaml of camera](https://github.com/flutter/plugins/blob/edfd9c0bf2602e3000f9689bc59431cf56fa99bc/packages/camera/camera/pubspec.yaml#L30) specifies version `^3.0.0`.

**Comment by mtrezza** [comment]
> pub publish wont upload the lock file. So we can't publish it.

If that is really a limitation of pub.dev, then that seems like a flaw to me. It basically means that the CI we run is almost irrelevant, because the developers using the Parse Flutter SDK may be using other dependency versions.

Yes, the caret means "no breaking change" upgrades, but any version upgrade (major, minor, patch) can potentially introduce a bug. That bug would be detected in our CI.

**Comment by fischerscode** [comment]
> If that is really a limitation of pub.dev, then that seems like a flaw to me.

You can [download](https://pub.dev/packages/camera/versions) the code uploaded to pub.dev. It doesn't contain the lock file.

> Yes, the caret means "no breaking change" upgrades, but any version upgrade (major, minor or patch) can potentially introduce a bug.

> Maybe we get in touch with pub.dev to inquire about this limitation?

I don't think they will change how pub is working, I think the team of google really thought about what is a good behavior.
And to be honest, it is possible to pin the versions by using `1.2.3` instead of `^1.2.3`, it's not just how most packages do it and how it's probably recommended.

Of course the approach of dynamically resolving the versions of the dependencies has a downside when it comes to the situation you described, but I think it is the best way they could have done it when they designed pub, since it allows the usage of irregularly updated packages while using patched nested dependencies.
So especially when it comes to security, I think it is a good approach.

If you feel like it would be better for the overall concept of the parse-platform, we could also introduce `*packagename*_fixed` packages, that have the dependencies fixed. But personally I think it should not be needed, since the dependencies we use (hopefully) have good testing. 

> I guess in the meantime we can just publish with caret to finally test our the release process.

👍

**Comment by mtrezza** [comment]
Quick update: I'll do a test release in the coming days so we can verify the manual release mechanism.

**Comment by mtrezza** [comment]
Test status:

- Dry run prevents publishing a version that already exists in the registry:
  > Local version: [3.1.0]
  > Remote version: [3.1.0]
  > Remote & Local versions are equal, skip publishing.

**Comment by mtrezza** [comment]
Regarding `pubspec.lock`, I think I understand now why the lock file is not part of the SDK in the flutter ecosystem:

- Dependency resolution is not as sophisticated as for example in npm, it can't handle if two dependencies require different versions of the same sub-dependency, see [docs](https://docs.flutter.dev/development/packages-and-plugins/using-packages#conflict-resolution).
- There are no dependency upgrade / vulnerability notification services in the Flutter ecosystem (at least not that I now of), so locking in a specific version creates too much inertia across the whole registry. Allowing "packages to evolve independently" as they they put it euphemistically is just a half-baked solution for that shortcoming. If a package fixes a vulnerability with a major version increment, the caret version range that is so popular in Flutter doesn't help much and the developer need to manually check whether a new major version is available and doesn't get any security notification.

**Comment by fischerscode** [comment]
@mtrezza Sorry for the late response...

> Dry run locally prevents publishing a version that already exists in the registry:

👍 

> Dry-run for flutter may not work as part of release workflow, because as long as dart dependency is not published, resolving fails; maybe resolving can be skipped for dry-run? Anyway, deactivating dry-run job for now.

Makes sens, forgot about that... Don't think there is a way around that. Since dry-run runs the test, as far as I know.

> Not sure how we can publish different versions for flutter / dart if GitHub allows only 1 version to be specified for a release. Since this is a fake monorepo, it may not even be possible without release-automation logic. For manual release it also wouldn't work to publish the dart and flutter independently, because the git tag history would be the same.

Maybe tag with `dart-vx.y.z` and  `flutter-vx.y.z` (or something similar)?

> Dart publishing seems to work, skipped for next release, as 3.1.1 is a flutter-only release. The following warning is reported:

No priority, since it is only the example, but I think it would be best to remove them from the .gitignore, since it seems to be suggested.

**Comment by mtrezza** [comment]
> No priority, since it is only the example, but I think it would be best to remove them from the .gitignore, since it seems to be suggested.

Do you know what these files are for? The path contains `ephemeral`, so it looks like temporary / cache files?

**Comment by mtrezza** [comment]
They look like ephemeral symlinks, so I don't think we should check them in. The symlinks are likely deployment specific.
-

**Comment by mtrezza** [comment]
I managed to run https://github.com/parse-community/Parse-SDK-Flutter/actions/runs/2404786674 successfully in dry-run. II think we are good now to make a release.

The GitHub Action will only publish the dart or flutter package if the local version is different from the currently published version, independently for each package. For example, if there has been only a change in the flutter package, it would publish only the flutter package, skipping the dart package. That is a strategy we can look into once we get to release automation and use an actual monorepo framwork. Until then we stick to the current approach of releasing both packages together with a common version number.

The question is how the release tag should be named, in this case we could name it something like `@flutter/

**Comment by mtrezza** [comment]
@fischerscode @RodrigoSMarques It seems we have successfully published [flutter-3.1.1](https://pub.dev/packages/parse_server_sdk_flutter).

Could you please check whether the published package is OK?

The release process currently works like this:

1. Merge PR that includes version increment and changelog entries for the specific package:
  - only add changelog entries to the package to which the changes relate (entries should already be added with every PR)
  - make sure changelog entries conform to Parse Platform org-wide format, see Parse Server changelog
  - only increment the version of a package if the package has changelog entries
  - increment the version according to the type of changes:
    - major version increment if there is a breaking change, or
    - minor version increment if there is at least one `feat` entry, or
    - patch version increment if there are only `fix` entires
  - make sure the comparison link at the top of the changelog has the correct tags
2. Create a GitHub release
  - Tag: `<package>-<version>`, e.g. `dart-1.2.3` or `flutter-1.2.3`
  - Description: Copy/paste the changelog entry from the package
3. Trigger release
  - Manually trigger the GitHub Action workflow `release-manual`
    - Tag: same as release tag, e.g. `dart-x.y.z`
    - Package: choose the package to relase, e.g. `dart `

Note:
- If a change spans both packages, the dart package has to be published first. As the flutter package depends on the dart package, publishing the flutter package that depends on an unpublished version of the dart package will fail and not publish to the registry.

**Comment by fischerscode** [comment]
> Could you please check whether the published package is OK?

Looks good to me 🚀 

> make sure changelog entries conform to Parse Platform org-wide format, see Parse Server changelog

When is the hash/link of the commit added? After a PR got merged?

> Manually trigger the GitHub Action workflow release-manual

Is it allowed for everyone having the permission to create a GitHub Release?

**Comment by mtrezza** [comment]
> When is the hash/link of the commit added? After a PR got merged?

Well, you are right, the changelog entry is usually auto-generated after the merge. As long as we're doing manual releases, we can simplify the changelog entry I guess and omit the commit hash.

> Is it allowed for everyone having the permission to create a GitHub Release?

Actually, this was more for self-documentation. Only the maintainer team can trigger releases, that is only me currently. GitHub doesn't provide a fine-grained permission model, so we are restricting write permissions only to the org core team for the moment. We are looking into making this more automated in the future. The process in this repo will be like with other repos in the org: 

1. PR submitted
2. PR author asks `@parse-community/android-sdk-review` for review
3. Members of that team review (the more the better)
4. Once all doubts are cleared, someone asks `@parse-community/android-sdk-maintenance` for merge

**Comment by mtrezza** [comment]
@parse-community/flutter-sdk Does anyone know of a dart-specific framework that is used to manage mono-repositories?

**Comment by fischerscode** [comment]
The [firebase packages](https://github.com/firebase/flutterfire) are managed by utilizing [melos](https://pub.dev/packages/melos) ([docs](https://docs.page/invertase/melos)).

Since I found it (https://github.com/parse-community/Parse-SDK-Flutter/issues/668#issuecomment-1114084351) I've played a bit with it. And to be honest, it seems perfect for managing monorepos without much work. As a bonus it supports linking the packages using a "path"-dependency for development without changing the `pubspec.yaml`.

**Comment by fischerscode** [comment]
While having [this](https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md) document in mind, I propose the following release procedure.

### Procedure:
1. PRs are created and merged without changes to the versions. (They use [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) as titles and are squash merged)
2. Once it is time to release, a authorized person triggers the workflow `release-preparation`.
3. The workflow runs `melos version --no-git-tag-version` (on master), tests, `melos publish` (it's a dry run), commits the changes (`chore(release): prepare for release`) and then creates a PR to master called `Release` (or something similar). The PR gets a label called `release`. (Additional tests run on only those PRs.)
4. Once the PR is approved and merged, CI (workflow `release`) triggers automatically and runs `melos publish --no-dry-run --git-tag-version`. Afterwards, the generated git tags are pushed.

*All names are of course only suggestions and can be changed.*

### Considerations:
- Do we want to support pre-releases of the packages (pub.dev supports it.)
- Do we want to publish after each commit or only when a human triggers it.

### Advantages:
- Due to the PR, nothing is published on accident.
- Little work when publishing.
- In case something unexpected happens, an admin could easily publish manually using the melos commands.
- All the boring stuff (versions, changelog) is managed by melos. 
- The pubspec.yaml's could use real version dependencies, since `melos bootstrap` links them locally.

### Disadvantages:
- Releases have to be triggered manually.
- Everybody contributing has to install melos (`dart pub global activate melos`) to bootstrap the repo.
- Due to melos changing the pubspec.lock for locally linking the packages of the repo, #750 might not be possible.

**Comment by mtrezza** [comment]
> Do we want to support pre-releases of the packages

Yes

> Do we want to publish after each commit or only when a human triggers it.

As we currently do it org wide; release an alpha version after each commit, once every month release a beta version with all changes of alpha, and a stable version with all changes in beta.

**Comment by mbfakourii** [comment]
@mtrezza 

At what stage are these issues? Is the problem solved?

**Comment by mtrezza** [comment]
There is no problem, auto-release is not just added yet because it needs more investigation. We have auto-release processes in other repos, but not for a mono-repository as this one.

**Comment by mtrezza** [comment]
After some tests with mono-repo release using semantic-release, it seems quite a shaky endeavor. semantic release don't support mono-repos OOTB and doesn't plan on doing so anytime soon for strategic reasons. Extensions ([multi-semantic-release](https://github.com/dhoulb/multi-semantic-release), [semantic-release-monorepo](https://github.com/pmowrer/semantic-release-monorepo) as at best hacks.

**Comment by Moumouls** [comment]
sounds promising @mtrezza i can suggest to take a look to Turbo repo/Changeset ( combined with npm or yarn)

https://turbo.build/repo/docs/handbook/publishing-packages/versioning-and-publishing

**Comment by Nidal-Bakir** [comment]
@mtrezza 
I'm really sorry but at this point, I'm lost there are `72 comments`. Can you give me a link or something to read in order to understand what is our end goal? I really want to help and work on this but until now I did not get the full pication of what is the issue or what is the current limitation.

Thanks.

**Comment by mtrezza** [comment]
The current CD is not automated. I have to manually make the release and the PR author has to manually write the changelog.

**Comment by mbfakourii** [comment]
@mtrezza 
Is this useful for these issues?
https://dart.dev/tools/pub/automated-publishing

**Comment by mtrezza** [comment]
We are already publishing to pub.dev automated. But I need to trigger this manually and we need to write the changelog manually.

**Comment by mtrezza** [comment]
Could someone answer this: Since the Parse Flutter SDK has a dependency on the Parse Dart SDK, would it make sense to automatically publish a new Flutter SDK version when a new Dart SDK version has been published? So that the Flutter SDK always depends on the latest Dart SDK?

**Comment by mtrezza** [comment]
Remaining issue is that the pub dev publish workflow is not automatically triggered, so push is done via local CLI. Opened a GitHub ticket for that.

**Comment by mtrezza** [comment]
https://github.com/parse-community/Parse-SDK-Flutter/pull/1092 should fix the auto publishing to pub; if this is validated in next release, this issue can be closed as release automation will be fully implemented.

**Comment by mtrezza** [comment]
Auto publishing works, but release workflow is triggered again; optimize for not running when commit message begins with chore(release).
